### PR TITLE
[AIRFLOW-3932] Optionally skip dag discovery heuristic.

### PR DESCRIPTION
**Pull request by jmcarp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3932
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by XD-DENG** [comment]
Hi @jmcarp , may you advise what's the use case for this? Are there DAGs without 'DAG'/'airflow' inside?

**Comment by jmcarp** [comment]
@XD-DENG: yes, I described this in more detail in https://issues.apache.org/jira/browse/AIRFLOW-3932, but in some cases I define a function that creates a dag (or multiple dags) in a shared module and import it in multiple dag files. If those helper functions are named `create_dag` or `create_dags`, the dag files might include the string `DAG` and will be ignored by the dagbag. Here's an example:

```python
# helpers.py
from airflow import DAG

def create_dag_for_table(table):
    ...

# dags.py
from .helpers import create_dag_for_table

dag = create_dag_for_table(my_table)
```

**Comment by XD-DENG** [comment]
Get your point.

But the concern is that whatever files in `dag_folder` will be processed for scheduling if you disable discovery heuristic, which may be very expensive. In your `dags.py`, you can simply add 'DAG'/'airflow' into it (adding as comment will do as well). This may be a better workaround for your use case?

**Comment by jmcarp** [comment]
I think it's clearer and more explicit to ignore files using `.airflowignore` than to silently drop files based on string contents. Ignoring with `.airflowignore` should also be slightly faster than the heuristic, since we don't have to read each file and check each line for matching strings. Including "DAG" and "airflow" in a comment works, but if a developer forgets to do it, they'll be confused and wonder why their DAG isn't showing up. If we want to whitelist specific files, I think it would make more sense to do it explicitly, by defining a list of files to load, rather than implicitly via string contents.

Anyway, to minimize any disruption, the `use_dag_discovery_heuristic` flag defaults to `True`, so users would only opt in if the heuristic is causing headaches for them. What do you think?

**Comment by XD-DENG** [comment]
Then from another perspective,  the developer may forget to add the files that should be ignored into `.airflowignore` ;-)

**Comment by jmcarp** [comment]
That seems like a reasonable choice to offer users. If some users want explicit control over dag discovery without workarounds and are willing to maintain an `.airflowignore`, they can set `use_dag_discovery_heuristic` to `False`. Users who want to use the heuristic (or don't care either way) can use the default of `True`.

cc @ashb @feng-tao and anybody else who might have an opinion.

**Comment by ashb** [comment]
I've had to work around this in the paste. The entire contents of one of my dag files was:

```
# Import packaged DAG. Do not remove this comment. Airflow needs to see "DAG" in the file
from mycompany.airflow.dags.terminate_idle_emr_clusters import dag, task  # noqa
```

File: airflow/utils/dag_processing.py [hunk 1]
```diff
@@ -332,17 +332,18 @@ def list_py_file_paths(directory, safe_mode=True,
...
-                    might_contain_dag = True
-                    if safe_mode and not zipfile.is_zipfile(file_path):
-                        with open(file_path, 'rb') as f:
-                            content = f.read()
-                            might_contain_dag = all(
-                                [s in content for s in (b'DAG', b'airflow')])
-
-                    if not might_contain_dag:
-                        continue
+                    if conf.getboolean('core', 'USE_DAG_DISCOVERY_HEURISTIC'):
```
Comment (ashb) [inline airflow/utils/dag_processing.py]: Yes, except not like this. Pass the value of this config option in to the pre-existing `safe_mode` parameter:

```
    :param safe_mode: whether to use a heuristic to determine whether a file
        contains Airflow DAG definitions
```

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by jmcarp** [comment]
Thanks @ashb. I updated the patch, and changed the name of the new configuration variable to match the existing flag.

**Comment by feng-tao** [comment]
lgtm, didn't know the safe_mode param before.

**Comment by feng-tao** [comment]
given you add a new config, could you update the UPDATE.md to include that new config information?

**Comment by jmcarp** [comment]
Thanks @feng-tao, I updated the docs.

**Comment by feng-tao** [comment]
lgtm

**Pull request merged**